### PR TITLE
SynthVAE Version 2

### PR DESCRIPTION
**Pull request by DaveBrind**
**Additions**
- GMM Pre-Processing Fix
- Hyperparameter Tuning
- Loss visualisation & saving
- Support for internal MIMIC sets
- Notebook implementation for hyperparameter tuning & Investigations

**Still To Be Done**
- Clean up documentation files (requirements.txt to new library versions) - Requirments.txt is now complete with updated versioning
- Clean up & check the old codes sdv_baselines.py, scratch_vae_expts.py - All code has been formatted using black and Dom's old codes have been updated
- Bring in documentation showing how to get GPU support - GPU support section is now in the readme
- Update Changelog.md - Still need to update

**Notable File Changes**

- Brought in RDT but only change two sections of code within the transformers folder -> numerical.py -> random_state added in init function line 582, this is then used as argument at line 626 when defining GMM model
- Folders investigations & Hyperparameter Tuning just contain code for training the model (with or without dp) as well as plot and calculate statistical & privacy metrics
- sdv_baselines, scratch_vae_expts have had parse args updated with new features
- Both Opacus & RDT have been reformatted using black so a lot of line changes will be due to this

### CODE REVIEW

**Comment by JRPearson500** [comment]
Could you add a link to MIMIC notebook for the "follows the data guidance provided in the Github repository"

**Comment by danjscho** [comment]
Thanks for the changes so far - can you make all the filenames have no spaces as well, for the same reasons as discussed (replacement with underscores is fine).
e.g. `Hyperparameter_Tuning/Hyperparameter Tuning MIMIC.py` -> `Hyperparameter_Tuning/Hyperparameter_Tuning_MIMIC.py`

File: README.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 ### About the Project
 
-This repository holds code for the NHSX Analytics Unit PhD internship project (previously known as Synthetic Data Generation - VAE) contextualising and investigating the potential use of Variational AutoEncoders (VAEs) for synthetic health data generation undertaken by Dominic Danks.
+This repository holds code for the NHSX Analytics Unit PhD internship project (previously known as Synthetic Data Generation - VAE) contextualising and investigating the potential use of Variational AutoEncoders (VAEs) for synthetic health data generation initially undertaken by Dominic Danks (last commit to the repository: commit 88a4bdf). This has then been further extended through the work undertaken by David Brind.
```
Comment (danjscho) [inline README.md:6]: Add URL link to commit.

File: README.md [hunk 2]
```diff
@@ -14,15 +14,16 @@ _**Note:** No data, public or private are shared in this repository._
 - The main code is found in the root of the repository (see Usage below for more information)
 - The accompanying [report](./reports/report.pdf) is also available in the `reports` folder
 - More information about the VAE with Differential Privacy can be found in the [model card](./model_card.md)
+- scratch_vae_expts.py is similar to the files you will find within the investigations folder. To re-run our results then scratch_vae_expts.py is all you require. If you want an easy way to understand our code and work process then using the respective notebooks within the investigations folder helps to run through the work.
```
Comment (danjscho) [inline README.md:17]: Any mention of file paths need wrapping in \` e.g. scratch_vae_expts.py -> `scratch_vae_expts.py`

File: README.md [hunk 3]
```diff
@@ -14,15 +14,16 @@ _**Note:** No data, public or private are shared in this repository._
...
-**N.B.** A modified copy of [Opacus](https://github.com/pytorch/opacus) (v0.14.0), a library for training PyTorch models with differential privacy, is contained within the repository. See the [model card](./model_card.md) for more details.
+**N.B.** A modified copy of [Opacus](https://github.com/pytorch/opacus) (v0.14.0), a library for training PyTorch models with differential privacy, is contained within the repository. A modified copy of [RDT](https://github.com/sdv-dev/RDT) (v0.6.2) that allows us to set the <b>random_state</b> for our GMM transformer method is also contained within the repository. See the [model card](./model_card.md) for more details.
 
 ### Built With
 
-[![Python v3.8](https://img.shields.io/badge/python-v3.8-blue.svg)](https://www.python.org/downloads/release/python-380/)
-- [PyTorch](https://github.com/pytorch)
-- [SDV](https://github.com/sdv-dev/SDV)
-- [Opacus](https://github.com/pytorch/opacus)
+[![Python v3.8.12](https://img.shields.io/badge/python-v3.8-blue.svg)](https://www.python.org/downloads/release/python-380/)
```
Comment (danjscho) [inline README.md:23]: No need to include `.12` beyond `3.8` for python version as should be compatible

File: README.md [hunk 4]
```diff
@@ -38,23 +39,48 @@ To create a suitable environment:
...
+
+We use:
+- `'pip install torch==1.9.0+cu111 torchvision==0.10.0+cu111 torchaudio==0.9.0 -f https://download.pytorch.org/whl/torch_stable.html'` 
+after using the pip install on requirements file in order to get the gpu supported versions of PyTorch that we require.
+
+Note that the model used is a simple MLP encoder/decoder and we are working with tabular data. Therefore the speed up provided by GPU is not that noticeable and it is easy to train using the CPU version for people without GPU support.
 
 ### Usage
 
+There are three main sections of interest (2 folders and a collation of python files): Hyperparameter tuning, investigations and files containing sdv_baselines, scratch_vae_expts and plot. Hyperparameter tuning and investigations can use either SUPPORT or internal MIMIC datasets. sdv_baselines, scratch_vae_expts and plot all use SUPPORT only. If you want to run these quickly then below outlines command line implementation.
```
Comment (danjscho) [inline README.md:62]: Match new folder names here and use \` when referencing folders specifically

File: README.md [hunk 5]
```diff
@@ -38,23 +39,48 @@ To create a suitable environment:
...
 
 optional arguments:
   -h, --help            show this help message and exit
   --n_runs N_RUNS       set number of runs/seeds
-  --model_type {CopulaGAN,CTGAN,GaussianCopula,TVAE}
+  --model_type          {CopulaGAN,CTGAN,GaussianCopula,TVAE}
                         set model for baseline experiment
+  --pre_proc_method     {GMM, standard}
+                        set the pre-processing method
+  --savemetrics         {True, False}
```
Comment (danjscho) [inline README.md:80]: `savemetrics` flag -> `save_metrics` to align and fix in code

File: README.md [hunk 6]
```diff
@@ -64,14 +90,23 @@ To reproduce the experiments contained in the report involving the VAE with/with
...
-usage: scratch_vae_expts.py [-h] [--n_runs N_RUNS] [--diff_priv DIFF_PRIV] [--savefile SAVEFILE]
+usage: scratch_vae_expts.py [-h] [--n_runs N_RUNS] [--diff_priv DIFF_PRIV] [--savefile SAVEFILE] [--savevisualisation {True, False}] [--savemetrics {True, False}] [--pre_proc_method {GMM, standard}] [--gower {True, False}]
 
 optional arguments:
   -h, --help            show this help message and exit
   --n_runs N_RUNS       set number of runs/seeds
   --diff_priv DIFF_PRIV
                         run VAE with differential privacy
+  --n_epochs            number of epochs to train for (early stopping is used by default)
   --savefile SAVEFILE   save trained model's state_dict to file
```
Comment (danjscho) [inline README.md:101]: Add `_` for all instance here: e.g. `savefile` -> `save_file` etc.

File: README.md [hunk 7]
```diff
@@ -85,11 +120,62 @@ usage: plot.py [-h] --savefile SAVEFILE
 optional arguments:
   -h, --help           show this help message and exit
   --savefile SAVEFILE  load trained model's state_dict from file
```
Comment (danjscho) [inline README.md:122]: Same as above

File: README.md [hunk 8]
```diff
@@ -85,11 +120,62 @@ usage: plot.py [-h] --savefile SAVEFILE
...
+  --pre_proc_method    {GMM, standard}
+                       specify the pre-processing method that you wish to employ
 ```
 
+#### Outputs Of The Model
+
+There are a selection of plots & metrics the model can output. These are given by parse arguments <b>--savemetrics</b>, <b>--savevisualisation</b> or functions <b>plot_elbo</b>, <b>plot_likelihood_breakdown</b>, <b>plot_variable_distributions</b> & <b>distribution_metrics</b> in the notebooks provided. These outputs give you a graph displaying the ELBO breakdown during training, the breakdown of categorical & numerical likelihoods, a comparison of variable distributions between original data & synthetic data for each variable as well as a csv file displaying all the distributional metrics from SDV.
```
Comment (danjscho) [inline README.md:129]: flags and names wrapped in \` e.g. <b>--savemetrics</b> -> `--save_metrics` (aligning with previous comment)

File: README.md [hunk 9]
```diff
@@ -85,11 +120,62 @@ usage: plot.py [-h] --savefile SAVEFILE
...
+| SVCDetection | GMLogLikelihood | CSTest | KSTestExtended | KSTestExtended | ContinuousKLDivergence | DiscreteKLDivergence |
+| --- | --- | --- | --- | --- | --- | --- |
+| 0.32 | -520.01 | 0.91 | 0.85 | 0.87 | 0.91 | 0.97 |
+| 0.31 | -523.21 | 0.90 | 0.86 | 0.88 | 0.92 | 0.99 |
+
+For information on these metrics then look in the [report](./reports/report.pdf) as well as the [SDV single table metrics documentation](https://sdv.dev/SDV/user_guides/evaluation/single_table_metrics.html)
+
+#### Note On Reproducibility Of Results
+
+In order to get reproducible results we have added in the <b>random_state</b> argument to the RDT transformers in order to set the sklearn's <b>random_state</b> argument. This results in the GMM pre-processing method producing the same transformation each run for the same dataset.
```
Comment (danjscho) [inline README.md:172]: <b>random_state</b> -> `random_state`

**Review by danjscho — COMMENTED** [review]
Some minor changes to README.md (and bits of codebase) suggested.
I also suggest running `black-nb` on the notebooks (https://pypi.org/project/black-nb/) for consistency with the rest of the codebase.

File: Hyperparameter_Tuning/MIMIC.ipynb [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# MIMIC Hyperparameter Tuning\n",
+    "\n",
+    "This notebook runs through hyperparameter tuning for the internal MIMIC dataset. For this we use the Optuna library.\n",
+    "\n",
+    "The notebook that produces our single table is found here [https://github.com/DaveBrind/SynthVAE/blob/main/MIMIC_preproc.ipynb]. If you want to create a single table yourself then follow the example csv file given at [https://github.com/DaveBrind/SynthVAE/blob/main/example_csv.csv]\n",
```
Comment (danjscho) [inline Hyperparameter_Tuning/MIMIC.ipynb:11]: Wrap URLs in `<...>` to render them explicitly in markdown rather than `[...]`

File: Hyperparameter_Tuning/SUPPORT.ipynb [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+    "- We are fitting a model to transform the data and hence we are approximating before model is trained. This will lose fidelity as the distribution will not be transformed perfectly.\n",
+    "\n",
+    "SUPPORT is a limited dataset as it has no missingness (which our model currently does NOT handle) and it has no datetime columns or other data types. Be wary drawing any conclusions from this set due to these constraints as well as the dataset size. Testing/training new models with this set could be useful but conclusive results should be tested on other sets."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
```
Comment (danjscho) [inline Hyperparameter_Tuning/SUPPORT.ipynb:111]: Some outputs have been committed here

File: VAE.py [hunk 12]
```diff
@@ -16,18 +18,11 @@ class Encoder(nn.Module):
     """
 
     def __init__(
-        self,
-        input_dim,
-        latent_dim,
-        hidden_dim=32,
-        activation=nn.Tanh,
-        device="cpu",
+        self, input_dim, latent_dim, hidden_dim=32, activation=nn.Tanh, device="gpu",
```
Comment (danjscho) [inline VAE.py:21]: It seems like it would be good to add flexibility to codebase to allow for hidden layers to be passed in a list/tuple e.g. (128, 64) c.f. `sdv` such that we can have multiple hidden layers of different sizes.
This would lead to a setup like 128 -> 64 -> 32 (latent) -> 64 -> 128.
Might be for a follow up PR - maybe add as an issue?

File: Investigations/SUPPORT_Notebook.ipynb [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+    "\n",
+    "metrics = distribution_metrics(\n",
+    "    gower_bool=gower,\n",
+    "    data_supp=data_supp,\n",
+    "    synthetic_supp=synthetic_supp,\n",
+    "    categorical_columns=original_categorical_columns,\n",
+    "    continuous_columns=original_continuous_columns,\n",
+    "    saving_filepath=None,\n",
+    "    pre_proc_method=pre_proc_method,\n",
+    ")\n"
```
Comment (danjscho) [inline Investigations/SUPPORT_Notebook.ipynb:382]: Possibly add a print of the metrics here to show the outputs in the notebook as well

File: Investigations/SUPPORT_Notebook.ipynb [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+    "\n",
+    "privacy_metric = privacy_metrics(\n",
+    "    private_variable=private_variable,\n",
+    "    data_supp=data_supp,\n",
+    "    synthetic_supp=synthetic_supp,\n",
+    "    categorical_columns=original_categorical_columns,\n",
+    "    continuous_columns=original_continuous_columns,\n",
+    "    saving_filepath=None,\n",
+    "    pre_proc_method=pre_proc_method,\n",
+    ")\n",
```
Comment (danjscho) [inline Investigations/SUPPORT_Notebook.ipynb:414]: Possibly add a print of the metrics here to show the outputs in the notebook as well

File: Investigations/MIMIC_Notebook.ipynb [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+    "metrics = distribution_metrics(\n",
+    "    gower_bool=gower,\n",
+    "    data_supp=original_metric_set,\n",
+    "    synthetic_supp=metric_synthetic_supp,\n",
+    "    categorical_columns=original_categorical_columns,\n",
+    "    continuous_columns=original_continuous_columns,\n",
+    "    saving_filepath=\"\",\n",
+    "    pre_proc_method=pre_proc_method,\n",
+    ")\n"
+   ]
```
Comment (danjscho) [inline Investigations/MIMIC_Notebook.ipynb:474]: Possibly add a print of the metrics here to show the outputs in the notebook as well

File: Investigations/MIMIC_Notebook.ipynb [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+    "private_variable = \"x14\"\n",
+    "\n",
+    "privacy_metric = privacy_metrics(\n",
+    "    private_variable=private_variable,\n",
+    "    data_supp=data_supp,\n",
+    "    synthetic_supp=synthetic_supp,\n",
+    "    categorical_columns=original_categorical_columns,\n",
+    "    continuous_columns=original_continuous_columns,\n",
+    "    saving_filepath=None,\n",
+    "    pre_proc_method=pre_proc_method,\n",
```
Comment (danjscho) [inline Investigations/MIMIC_Notebook.ipynb:504]: Possibly add a print of the metrics here to show the outputs in the notebook as well

File: Investigations/MIMIC_Notebook.ipynb [hunk 17]
```diff
@@ -0,0 +1,531 @@
...
+   "id": "c082cd4d",
+   "metadata": {},
+   "source": [
+    "# MIMIC Notebook\n",
+    "\n",
+    "This notebook runs through investigations on internal NHSX datasets collated from the MIMIC-III dataset.\n",
+    "\n",
+    "For users who do not have MIMIC-III access then investigations cannot be run through until access is completed. In the meantime you can access the investigations on the open access SUPPORT dataset.\n",
+    "\n",
+    "The notebook that produces our single table is found here [https://github.com/DaveBrind/SynthVAE/blob/main/MIMIC_preproc.ipynb]. If you want to create a single table yourself then follow the example csv file given at [https://github.com/DaveBrind/SynthVAE/blob/main/example_csv.csv]"
```
Comment (danjscho) [inline Investigations/MIMIC_Notebook.ipynb:14]: Fix `[...]` -> `<...>` for URLs

File: example_csv.csv [hunk 18]
```diff
@@ -0,0 +1,7 @@
+Categorical 1,Continuous 1,Datetime 1
```
Comment (danjscho) [inline example_csv.csv:1]: Could you rename the file to something like `example_input.csv` so that users don't mix up the `..._csv.csv` bit

Comment (danjscho) [reply example_csv.csv:1]: Will need updating where referenced as well.

File: .gitignore [hunk 19]
```diff
@@ -4,3 +4,4 @@ __pycache__
 *.pt
 *.pkl
 .DS_Store
+.ipynb_checkpoints
```
Comment (danjscho) [inline .gitignore:7]: Add `data` folder to ignore

File: Hyperparameter_Tuning/Hyperparameter_Tuning_MIMIC.py [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+original_categorical_columns = [
+    "ETHNICITY",
+    "DISCHARGE_LOCATION",
+    "GENDER",
+    "FIRST_CAREUNIT",
+    "VALUEUOM",
+    "LABEL",
+]
+original_continuous_columns = ["Unnamed: 0", "ROW_ID", "SUBJECT_ID", "VALUE", "age"]
```
Comment (danjscho) [inline Hyperparameter_Tuning/Hyperparameter_Tuning_MIMIC.py:59]: `"Unnamed: 0"` is coming from `index=False` missing in the writing out of the pandas dataframe from the MIMIC prepro script and is thus not useful.

**Review by danjscho — APPROVED** [review]
